### PR TITLE
Cosipbft: recover from a block propagation failure

### PR DESCRIPTION
**Pull request by Gilthoniel**
This improves the round procedure to handle cases where a block is
committed but the leader failed to propagate it to followers.

### CODE REVIEW

**Comment by Gilthoniel** [comment]
The address implementation of Minogrpc had to move to the session package because of cycling imports.

File: core/ordering/cosipbft/pbft/mod.go
Comment (nkcr) [inline core/ordering/cosipbft/pbft/mod.go:232]: then what happens to the block? The leader will wait for this block to be committed, which will reach a timeout?

File: core/ordering/cosipbft/pbft/mod.go
Comment (nkcr) [inline core/ordering/cosipbft/pbft/mod.go:254]: need an error message update

File: mino/minoch/mod.go
Comment (nkcr) [inline mino/minoch/mod.go:17]: isn't it a bit expensive to add all this mechanism just for the sake of testing ?

File: mino/minogrpc/session/addr.go
Comment (nkcr) [inline mino/minogrpc/session/addr.go:46]: we actually never check that a.host is valid

File: mino/minogrpc/session/addr.go
Comment (nkcr) [inline mino/minogrpc/session/addr.go:141]: it panics if len(text) == 1

File: mino/minogrpc/session/addr.go
Comment (Gilthoniel) [reply mino/minogrpc/session/addr.go:46]: Well if that's the case, it will fail to dial.

File: mino/minoch/mod.go
Comment (Gilthoniel) [reply mino/minoch/mod.go:17]: Minoch has been implemented for the sake of testing, so I don't think so. It's also very cheap as the default behaviour of being empty does add anything to the logic.

File: core/ordering/cosipbft/pbft/mod.go
Comment (Gilthoniel) [reply core/ordering/cosipbft/pbft/mod.go:232]: Yes, if the leader decides to propose a different block, the followers will simply reject it and proceed to a view change after the timeout. Then new leader == new block.

File: mino/minogrpc/session/addr.go
Comment (Gilthoniel) [reply mino/minogrpc/session/addr.go:141]: it actually doesn't: https://play.golang.org/p/LYMPrve_bQK

`[2:]` would panic though.

Comment (Gilthoniel) [reply mino/minogrpc/session/addr.go:141]: I'll add a test anyway.

**Review by nkcr — APPROVED** [review]

**Comment by nkcr** [comment]
Can be merged once the conflict is resolved.

**Comment by Gilthoniel** [comment]
I'll wait for #125 to be merged, otherwise it will be pain to rebase it.

**Pull request merged**
